File: movil/lib/models/incidente.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

part 'incidente.g.dart';

enum EstadoIncidente {
  disponible,
  @JsonValue('no_disponible')
  noDisponible,
}

@JsonSerializable()
class Incidente extends Equatable {
  final int id;
  final String descripcion;
  @JsonKey(name: 'fecha_hora')
  final DateTime fechaHora;
  final double latitud;
  final double longitud;
  final EstadoIncidente estado;
  @JsonKey(name: 'resumen_ia')
  final String? resumenIa;
  @JsonKey(name: 'cliente_id')
  final int clienteId;
  @JsonKey(name: 'vehiculo_id')
  final int? vehiculoId;
  @JsonKey(name: 'categoria_id')
  final int? categoriaId;

  const Incidente({
    required this.id,
    required this.descripcion,
    required this.fechaHora,
    required this.latitud,
    required this.longitud,
    required this.estado,
    this.resumenIa,
    required this.clienteId,
    this.vehiculoId,
    this.categoriaId,
  });

  factory Incidente.fromJson(Map<String, dynamic> json) =>
      _$IncidenteFromJson(json);

  Map<String, dynamic> toJson() => _$IncidenteToJson(this);

  @override
  List<Object?> get props =>
      [id, descripcion, fechaHora, latitud, longitud, estado, clienteId];
}

File: backend/app/models/incidente.py
import enum
from datetime import datetime
from sqlalchemy import Column, Integer, String, Float, DateTime, ForeignKey, Text, Enum as SAEnum
from sqlalchemy.orm import relationship
from app.models.base import Base


class EstadoIncidente(str, enum.Enum):
    disponible = "disponible"
    no_disponible = "no_disponible"


class Incidente(Base):
    __tablename__ = "incidentes"

    id          = Column(Integer, primary_key=True, index=True)
    descripcion = Column(Text, nullable=False)
    fecha_hora  = Column(DateTime, default=datetime.utcnow, nullable=False)
    latitud     = Column(Float, nullable=False)
    longitud    = Column(Float, nullable=False)
    estado      = Column(SAEnum(EstadoIncidente), default=EstadoIncidente.disponible, nullable=False)
    resumen_ia  = Column(Text, nullable=True)   # resumenIA del diagrama
    foto_incidente = Column(String(500), nullable=True)

    # --- Foreign Keys ---
    cliente_id  = Column(Integer, ForeignKey("clientes.id",  ondelete="CASCADE"),  nullable=False)
    vehiculo_id = Column(Integer, ForeignKey("vehiculos.id", ondelete="SET NULL"), nullable=True)
    categoria_id = Column(Integer, ForeignKey("categorias.id", ondelete="SET NULL"), nullable=True)

    # --- Relaciones ---
    cliente  = relationship("Cliente",  back_populates="incidentes")
    vehiculo = relationship("Vehiculo", back_populates="incidentes")
    categoria = relationship("Categoria", back_populates="incidentes")
    asignacion = relationship("AsignacionServicio", back_populates="incidente", uselist=False)
    evidencias = relationship("Evidencia", back_populates="incidente")
    # procesamientos_ia = relationship("ProcesamientoIA", back_populates="incidente")
